dependency-specification.rst

```diff
@@ -0,0 +1,324 @@
+acceptable, so the language permits describing all these cases.
+
+The language defined is a compact line based format which is already in
+widespread use in pip requirements files, though we do not specify the command
+line option handling that those files permit. There is one caveat - the
+URL reference form, specified in PEP-440 [#pep440]_ is not actually
+implemented in pip, but since PEP-440 is accepted, we use that format rather
+than pip's current native format.
+
+Motivation
+==========
+
+Any specification in the Python packaging ecosystem that needs to consume
+lists of dependencies needs to build on an approved PEP for such, but
+PEP-426 [#pep426]_ is mostly aspirational - and there are already existing
+implementations of the dependency specification which we can instead adopt.
+The existing implementations are battle proven and user friendly, so adopting
+them is arguably much better than approving an aspirational, unconsumed, format.
+
+Specification
+=============
+
+Examples
+--------
+
+All features of the language shown with a name based lookup::
+
+    requests [security,tests] >= 2.8.1, == 2.8.* ; python_version < "2.7.10"
+
+A minimal URL based lookup::
+
+    pip @ https://github.com/pypa/pip/archive/1.3.1.zip#sha1=da9234ee9982d4bbb3c72346a6de940a148ea686
+
+Concepts
+--------
+
+A dependency specification always specifies a distribution name. It may
+include extras, which expand the dependencies of the named distribution to
+enable optional features. The version installed can be controlled using
+version limits, or giving the URL to a specific artifact to install. Finally
+the dependency can be made conditional using environment markers.
+
+Grammar
+-------
+
+We first cover the grammar briefly and then drill into the semantics of each
+section later.
+
+A distribution specification is written in ASCII text. We use ABNF [#abnf]_ to
+provide a precise grammar. The grammar covers just the specification itself.
+The expectation is that the specification will be embedded into a larger
+system which offers framing such as comments, multiple line support via
+continuations, or other such features.
+
+Versions may be specified according to the PEP-440 [#pep440]_ rules. (Note:
+URI is defined in std-66 [#std66]_::
+
+    version-cmp   = "<" / "<=" / "!=" / "==" / ">=" / ">"
+    version       = 1*( DIGIT / ALPHA / "-" / "_" / "." / "*" )
+    version-inner = version-cmp version *(',' version-cmp version)
+    versionspec   = ("(" version-inner ")") / version-inner
+    urlspec       = "@" URI
+
+Environment markers allow making a specification only take effect in some
+environments::
+
+    marker-op     = version-cmp / "in" / "not in"
+    python-str-c  = WSP / ALPHA / DIGIT / "(" / ")" / "." / "{" / "}" /
+                    "-" / "_" / "*"
+    python-str    = "'" *(python-str-c / DQUOTE) "'"
+    python-str    =/ DQUOTE *(python-str-c / "'") DQUOTE
+    marker-var    = python-str / "python_version" / "python_full_version" /
```

Comment:
I think restricting it has utility, particularly for the version specifiers because they are designed to have the "Variable" portion be on the LHS and reverse that is going to give inconsistent or bad results, particularly for ``~=`` and ``==1.*`` and such.

I don't think it makes either the grammar or implementation more complex in a meaningful way. You just need to define something like:

```
marker-var-lhs "python_version" / "python_full_version" /
                       "os_name"" / "sys_platform" / "platform_release" /
                       "platform_version" / "platform_machine" /
                       "platform_python_implementation" / "implementation_name" /
                       "implementation_version" / "platform_dist_name"
                       "platform_dist_version" / "platform_dist_id"
marker-var-rhs python-str
marker-expr   = "(" *WSP marker *WSP ")"
                      =/ (marker-var-lhs [*WSP marker-op *WSP marker-var-rhs])
```

I don't think having to take one variable and split it into two and call them both once instead of one two is a meaningful amount of difference in complexity.